### PR TITLE
Uikit Rework - move existing business components into front-kit (Product, Category, SignIn, ...)

### PR DESCRIPTION
**Pull request by tutok**
### Please check if the PR fulfills these requirements ###

- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
moved all existing (in `falcon-ecommerce-uikit` package) business components into new `@deity/falcon-front-kit` (except checkout)
 
**What is the current behavior? (You can also link to an open issue here)**

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change? (What changes might users need to make in their application due to this PR?)**

**Other information:**

### CODE REVIEW

File: packages/falcon-ui-kit/src/Product/ProductGalleryThumbLayout.tsx [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import React from 'react';
+import { Box, themed } from '@deity/falcon-ui';
+
+export const ProductGalleryThumbLayout = themed({
+  tag: Box,
+  defaultTheme: {
+    ProductGalleryThumbLayout: {
```
Comment (mateuszwozniak) [inline packages/falcon-ui-kit/src/Product/ProductGalleryThumbLayout.tsx]:
```suggestion
    productGalleryThumbLayout: {
```

File: packages/falcon-ui-kit/src/Product/ProductLayout.tsx [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  meta: 'meta',
+  empty: 'empty',
+  options: 'options',
+  error: 'error'
+};
+
+export const ProductLayout = themed({
+  tag: 'article',
+  defaultTheme: {
+    ProductLayout: {
```
Comment (mateuszwozniak) [inline packages/falcon-ui-kit/src/Product/ProductLayout.tsx]:
```suggestion
    productLayout: {
```

File: packages/falcon-ui-kit/src/Product/ProductOptionRadio.tsx [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import React from 'react';
+import { themed, Radio } from '@deity/falcon-ui';
+
+export const ProductOptionRadio = themed({
+  tag: Radio,
+  defaultTheme: {
+    ProductOptionRadio: {
```
Comment (mateuszwozniak) [inline packages/falcon-ui-kit/src/Product/ProductOptionRadio.tsx]:
```suggestion
    productOptionRadio: {
```

File: packages/falcon-front-kit/src/Forms/FormProvider.ts [hunk 4]
```diff
@@ -0,0 +1,6 @@
+import { FormikConfig, FormikValues } from 'formik';
+
+export type FormProviderProps<Values = FormikValues> = {
+  onSubmit?: () => any | Promise<any>;
```
[the remark this answers is not in the document]
Comment (tutok) [reply packages/falcon-front-kit/src/Forms/FormProvider.ts]: Yeah, I have had that dilemma too (https://github.com/deity-io/falcon/blob/130982411f0041188637e418827ddb215a104519/packages/falcon-front-kit/src/Forms/FormProvider.ts#L4), will change

File: packages/falcon-shop-extension/schema.graphql [hunk 5]
```diff
@@ -1,15 +1,15 @@
 extend type Query {
   menu: [MenuItem] @cache(ttl: 15)
-  category(id: String!): Category @cache(ttl: 15)
-  product(id: String!): Product @cache(ttl: 5)
-  products(
-    categoryId: String
+  category(id: ID!): Category @cache(ttl: 15)
+  product(id: ID!): Product @cache(ttl: 5)
+  productList(
```
[the remark this answers is not in the document]
Comment (tutok) [reply packages/falcon-shop-extension/schema.graphql]: > I'd assume it was done to reflect the returning type, but with GraphQL - we don't rely on type names that much (only query/mutation names matter)

File: examples/shop-with-blog/client/src/pages/Home.js [hunk 6]
```diff
@@ -25,14 +26,14 @@ const HOMEPAGE_PRODUCTS_QUERY = gql`
 `;
 
 const Home = () => (
-  <GridLayout gridGap="md" py="ld">
+  <PageLayout>
     <H1 css={{ textAlign: 'center' }}>
       <T id="home.hotSellers" />
     </H1>
     <Query query={HOMEPAGE_PRODUCTS_QUERY} variables={{ categoryId: '25', amount: 20 }}>
```
[the remark this answers is not in the document]
Comment (tutok) [reply examples/shop-with-blog/client/src/pages/Home.js]: I simply forget to replace `HOMEPAGE_PRODUCTS_QUERY` via the parametrized version of `GET_PRODUCT_LIST`

image

on the end, `HOMEPAGE_PRODUCTS_QUERY` will go into the bin

File: packages/falcon-shop-extension/schema.graphql (hunk 5 above)
[the remark this answers is not in the document]
Comment (Borales) [reply packages/falcon-shop-extension/schema.graphql]: @tutok yes, as discussed - we'll update the rest of the query names in #565

File: packages/falcon-front-kit/src/Forms/FormProvider.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (tutok) [reply packages/falcon-front-kit/src/Forms/FormProvider.ts]: done, but on branch https://github.com/deity-io/falcon/pull/565 (https://github.com/deity-io/falcon/pull/565/commits/cfd412a8b4f6266d52a0f54e0e70d55bd8527279) as it was easier

[the remark this answers is not in the document]
Comment (tutok) [reply packages/falcon-front-kit/src/Forms/FormProvider.ts]: done, but on branch https://github.com/deity-io/falcon/pull/565 (https://github.com/deity-io/falcon/pull/565/commits/cfd412a8b4f6266d52a0f54e0e70d55bd8527279) as it was easier

File: packages/falcon-ui-kit/src/Product/ProductPrice.tsx [hunk 7]
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { ThemedComponentProps } from '@deity/falcon-ui';
+import { Price } from '../Price';
+
+export type ProductPriceProps = {
```
[the remark this answers is not in the document]
Comment (tutok) [reply packages/falcon-ui-kit/src/Product/ProductPrice.tsx]: ok, removed `shop-extension` dependency from example project

File: packages/falcon-front-kit/src/Forms/getInputDefaultValidators.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
+import { inputTypeToDefaultValidatorsMapper } from './inputTypeToDefaultValidatorsMapper';
+import { requiredValidator, Validator } from './validators';
+
+export function getInputDefaultValidators(
```
[the remark this answers is not in the document]
Comment (tutok) [reply packages/falcon-front-kit/src/Forms/getInputDefaultValidators.ts]: done

File: packages/falcon-front-kit/src/Forms/getInputDefaultValidators.ts [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+  input: Pick<React.InputHTMLAttributes<HTMLInputElement>, 'type' | 'required'> & { validate?: Validator[] }
+): Validator[] {
+  const { type, required, validate } = input;
+
+  const anyValidatorsDefined = !!validate;
+  let result = anyValidatorsDefined ? validate : [];
+
+  if (required) {
+    result = [requiredValidator, ...result];
```
[the remark this answers is not in the document]
Comment (tutok) [reply packages/falcon-front-kit/src/Forms/getInputDefaultValidators.ts]: done

File: packages/falcon-front-kit/src/Forms/getInputDefaultValidators.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+  }
+
+  if (!anyValidatorsDefined) {
+    result = [...result, ...inputTypeToDefaultValidatorsMapper(type)];
```
[the remark this answers is not in the document]
Comment (tutok) [reply packages/falcon-front-kit/src/Forms/getInputDefaultValidators.ts]: done

File: packages/falcon-front-kit/src/Forms/validators.ts [hunk 11]
```diff
@@ -3,8 +3,15 @@ export type Validator = (value: string, label: string) => string | undefined;
 // TODO: when new i18n support is ready use it to translate validation messages
 
 const validEmailRegex = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
+const validPasswordRegex = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)/;
+
+export const requiredValidator: Validator = (value, label) => {
+  if (value === null || value === undefined || value === '' || (Array.isArray(value) && !(value as any[]).length)) {
```
[the remark this answers is not in the document]
Comment (tutok) [reply packages/falcon-front-kit/src/Forms/validators.ts]: done

File: packages/falcon-ui-kit/src/Product/ProductGalleryThumbLayout.tsx (hunk 1 above)
Comment (tutok) [reply packages/falcon-ui-kit/src/Product/ProductGalleryThumbLayout.tsx]: done

File: packages/falcon-ui-kit/src/Product/ProductLayout.tsx (hunk 2 above)
Comment (tutok) [reply packages/falcon-ui-kit/src/Product/ProductLayout.tsx]: done

File: packages/falcon-ui-kit/src/Product/ProductOptionRadio.tsx (hunk 3 above)
Comment (tutok) [reply packages/falcon-ui-kit/src/Product/ProductOptionRadio.tsx]: done

File: examples/shop-with-blog/client/src/theme.js [hunk 12]
```diff
@@ -16,6 +16,7 @@ import CheckCircle from 'react-feather/dist/icons/check-circle';
 import Eye from 'react-feather/dist/icons/eye';
 import EyeOff from 'react-feather/dist/icons/eye-off';
 import { createTheme } from '@deity/falcon-ui';
+import { DeityLogo, SignInIcon } from '@deity/falcon-ui-kit';
```
[the remark this answers is not in the document]
Comment (tutok) [reply examples/shop-with-blog/client/src/theme.js]: done

File: examples/shop-with-blog/client/src/pages/Home.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (tutok) [reply examples/shop-with-blog/client/src/pages/Home.js]: while filtering products via category is only possible when fetching "category with products" I used `CategoryWithProductsQuery` here

File: packages/falcon-shop-extension/schema.graphql [hunk 13]
```diff
@@ -471,10 +471,21 @@ type CartTotal {
...
   bundleOptions: [BundleOptionInput]
 }
 
+input ProductOptionInput {
+  optionId: Int!
+  value: Int!
+}
+
+input BundleOptionInput {
+  optionId: Int!
```
[the remark this answers is not in the document]
Comment (tutok) [reply packages/falcon-shop-extension/schema.graphql]: done

**Pull request merged**